### PR TITLE
Add no-server-header support for Web Socket implementations

### PR DESCRIPTION
**Pull request by iudeen**

### CODE REVIEW

**Comment by iudeen** [comment]
I am not sure if these features are needed at this point of time, but I went ahead and added as I was not able to conclusively find a reason why server headers were not added to Websocket protocols.

Note: Also, we can keep this in WIP till Websocket library provides an API to disable server header completely.

**Review by Kludex — CHANGES_REQUESTED** [review]
I'd prefer to not use the workaround on the `websockets` implementation 🤔

There's already an issue on `websockets` about this, and we already documented the limitation. Let's modify the note to specify that it does not work on `websockets` implementation and add the headers to `wsproto`.

**Comment by iudeen** [comment]
@Kludex, agreed will make the changes today! thanks for your feedback

**Comment by Kludex** [comment]
Thanks 🙏

**Review by iudeen — COMMENTED** [review]
implemented request changes

**Comment by iudeen** [comment]
## Test Results:

### Code:
```python
# app.py

from fastapi import FastAPI
from starlette.websockets import WebSocket

app = FastAPI()


@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    while True:
        data = await websocket.receive_text()
        await websocket.send_text(f"Message text was: {data}")
```

### `Websockets` implementation without `--no-server-header`

```shell
>uvicorn app:app
INFO:     Started server process [9848]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
```

```
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: bxNWrBsDyxZs3+p8aaNcwTzh6ac=
Sec-WebSocket-Extensions: permessage-deflate
date: Fri, 12 Aug 2022 09:26:02 GMT
server: uvicorn
```

### `Websockets` implementation with `--no-server-header`

```shell
>uvicorn app:app --no-server-header
INFO:     Started server process [9848]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
```

```
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: voUjj70rAcpXQJnB1yI+0lLZuGg=
Sec-WebSocket-Extensions: permessage-deflate
date: Fri, 12 Aug 2022 09:29:49 GMT
Server: Python/3.10 websockets/10.3
```

### `Wsproto` implementation with `--no-server-header`

```shell
>uvicorn app:app --no-server-header --ws wsproto
INFO:     Started server process [9848]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
```

```
HTTP/1.1 101
Upgrade: WebSocket
Connection: Upgrade
Sec-WebSocket-Accept: u1f5bQpPsn8/VPZAo5ynbrRRkag=
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits=15
date: Fri, 12 Aug 2022 17:33:05 GMT
```

### `Wsproto` implementation without `--no-server-header`

```shell
>uvicorn app:app  --ws wsproto
INFO:     Started server process [9848]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
```

```
HTTP/1.1 101
Upgrade: WebSocket
Connection: Upgrade
Sec-WebSocket-Accept: r0+014Ywn5SbmlY9tNSEPg6FRfg=
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits=15
date: Fri, 12 Aug 2022 17:34:04 GMT
server: uvicorn
```

**Comment by iudeen** [comment]
`--no-date-header` is not compatible with Websockets implementation. Added that also as a limitation in documentation.

File: uvicorn/protocols/websockets/wsproto_impl.py [hunk 1]
```diff
@@ -248,7 +249,15 @@ async def send(self, message):
...
-                extra_headers = message.get("headers", [])
+                headers = list(message.get("headers", [])) + self.default_headers
+                extra_headers = []
+                _added_names = []
+                for name, value in headers:
+                    if name.lower() in _added_names:
+                        continue
+                    _added_names.append(name.lower())
+                    extra_headers.append((name, value))
+
```
Comment (Kludex) [inline uvicorn/protocols/websockets/wsproto_impl.py]: Why not just this? 🤔 
```suggestion
                extra_headers = list(message.get("headers", [])) + self.default_headers
```

**Comment by Kludex** [comment]
We also need tests here, jfyk.

File: uvicorn/protocols/websockets/wsproto_impl.py (hunk 1 above)
Comment (iudeen) [reply uvicorn/protocols/websockets/wsproto_impl.py]: if the user passes custom Server header, it lead to an exception of duplicate header names. That's why added a deduplication logic

Comment (Kludex) [reply uvicorn/protocols/websockets/wsproto_impl.py]: What exception?

Comment (iudeen) [reply uvicorn/protocols/websockets/wsproto_impl.py]: In `websockets` implementation 

```python
  File "...\lib\site-packages\websockets\datastructures.py", line 119, in __getitem__
    raise MultipleValuesError(key)
websockets.datastructures.MultipleValuesError: 'Server'
```

In `wsproto` we get duplicate `Server` headers. 

This happens when you add 

```
@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept(headers=[(b"Server", b"this is my server")])
    while True:
        data = await websocket.receive_text()
        await websocket.send_text(f"Message text was: {data}")
```

If its okay to have multiple headers when using `wsproto`, I can remove the logic.

Comment (Kludex) [reply uvicorn/protocols/websockets/wsproto_impl.py]: This should be a bug on `websockets` side. 🤔

It's ok to have multiple headers with the same key, it's the behavior we have on `http` protocols.

Comment (iudeen) [reply uvicorn/protocols/websockets/wsproto_impl.py]: Oh, okay! will do the necessary changes. And we should add that custom `Server` header is not allowed for `websockets`.  Or should I keep the validation for Websockets for now?

Comment (Kludex) [reply uvicorn/protocols/websockets/wsproto_impl.py]: Let's not allow on `websockets`.

**Comment by iudeen** [comment]
Closing this request, as it was made from `master` branch of the forked repo. Sorry!

Please refer to https://github.com/encode/uvicorn/pull/1606 for further updates.

**Pull request closed without merging**